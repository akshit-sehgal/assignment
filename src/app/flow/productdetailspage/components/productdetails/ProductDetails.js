import React from 'react';
import PropTypes from 'prop-types';
import ProductImage from './ProductImage';
import ProductInfo from './ProductInfo';
import DetailsShimmer from '../../../../shared/components/shimmer/DetailsShimmer';
import './productdetails.scss';
import { DETAILS_LABEL } from './constants';

class ProductDetails extends React.PureComponent {
    componentDidMount() {
        const { fetchProductDetails, id, productDetails } = this.props;
        if (!productDetails || (productDetails && productDetails.id !== id)) {
            fetchProductDetails(id);
        }
        window.scrollTo(0, 0);
    }
    generateDetailsView = () => {
        const { productDetails } = this.props;
        if (!productDetails) return;
        return (
            <React.Fragment>
                <div className="pd__left-panel">
                    <ProductImage imgSrc={productDetails.img} />
                </div>
                <div className="pd__right-panel">
                    <ProductInfo {...productDetails} />
                </div>
            </React.Fragment>
        )
    }
    render() {
        const { isFetchingDetails } = this.props;
        return (
            <div className="product-details section-container">
                <div className="pd__heading">{DETAILS_LABEL.HEADING}</div>
                {isFetchingDetails && <DetailsShimmer />}
                {this.generateDetailsView()}
            </div>
        );
    }
}
ProductDetails.propTypes = {
    fetchProductDetails: PropTypes.func.isRequired,
    isFetchingDetails: PropTypes.bool.isRequired,
    id: PropTypes.string.isRequired,
    productDetails: PropTypes.shape({
        id: PropTypes.string.isRequired,
        img: PropTypes.string.isRequired,
        rating: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        price: PropTypes.string.isRequired,
        description: PropTypes.string.isRequired
    })
}
export default ProductDetails;

