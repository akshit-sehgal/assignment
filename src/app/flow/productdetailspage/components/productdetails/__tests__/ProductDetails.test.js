import React from 'react';
import { shallow } from 'enzyme';
import ProductDetails from '../ProductDetails';
import { mockProductDetails } from '../__mocks__/productDetails.mock';

describe('Testing <ProductDetails />', () => {
    let props;
    window.scrollTo = jest.fn();
    beforeEach(() => {
        props = {
            id: '1',
            productDetails: mockProductDetails,
            isFetchingDetails: false,
            fetchProductDetails: jest.fn()
        }
    });
    it('should render the default view', () => {
        const wrapper = shallow(<ProductDetails {...props} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ProductImage').length).toEqual(1);
        expect(wrapper.find('ProductInfo').length).toEqual(1);
        wrapper.setProps({
            productDetails: null,
            isFetchingDetails: true
        });
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('ProductImage').length).toEqual(0);
        expect(wrapper.find('ProductInfo').length).toEqual(0);
        expect(wrapper.find('DetailsShimmer').length).toEqual(1);
    });
    it('should call fetchProductDetails on mount', () => {
        props.id = '2';
        const wrapper = shallow(<ProductDetails {...props} />);
        expect(props.fetchProductDetails).toHaveBeenCalledWith(props.id);
        expect(window.scrollTo).toHaveBeenCalledWith(0, 0);
    });

});