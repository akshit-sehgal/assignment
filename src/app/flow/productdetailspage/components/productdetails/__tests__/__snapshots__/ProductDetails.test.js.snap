// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ProductDetails /> should render the default view 1`] = `
<div
  className="product-details section-container"
>
  <div
    className="pd__heading"
  >
    Product Details
  </div>
  <div
    className="pd__left-panel"
  >
    <ProductImage
      imgSrc="https://images.google.com/1.png"
    />
  </div>
  <div
    className="pd__right-panel"
  >
    <ProductInfo
      description="Product Description"
      id="1"
      img="https://images.google.com/1.png"
      name="Product 1"
      price="Rs 2000"
      rating={4}
    />
  </div>
</div>
`;

exports[`Testing <ProductDetails /> should render the default view 2`] = `
<div
  className="product-details section-container"
>
  <div
    className="pd__heading"
  >
    Product Details
  </div>
  <DetailsShimmer />
</div>
`;
